com.unity.netcode.gameobjects/Runtime/Messaging/MessagingSystem.cs

```diff
@@ -0,0 +1,546 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using Unity.Collections;
+using Unity.Collections.LowLevel.Unsafe;
+using UnityEngine;
+
+namespace Unity.Netcode
+{
+
+    public class InvalidMessageStructureException : SystemException
+    {
+        public InvalidMessageStructureException() { }
+        public InvalidMessageStructureException(string issue) : base(issue) { }
+    }
+
+    public class MessagingSystem : IDisposable
+    {
+        private struct ReceiveQueueItem
+        {
+            public FastBufferReader Reader;
+            public MessageHeader Header;
+            public ulong SenderId;
+            public float Timestamp;
+        }
+
+        private struct SendQueueItem
+        {
+            public BatchHeader BatchHeader;
+            public FastBufferWriter Writer;
+            public readonly NetworkDelivery NetworkDelivery;
+
+            public SendQueueItem(NetworkDelivery delivery, int writerSize, Allocator writerAllocator, int maxWriterSize = -1)
+            {
+                Writer = new FastBufferWriter(writerSize, writerAllocator, maxWriterSize);
+                NetworkDelivery = delivery;
+                BatchHeader = default;
+            }
+        }
+
+        internal delegate void MessageHandler(ref FastBufferReader reader, NetworkContext context);
+
+        private NativeList<ReceiveQueueItem> m_IncomingMessageQueue = new NativeList<ReceiveQueueItem>(16, Allocator.Persistent);
+
+        private MessageHandler[] m_MessageHandlers = new MessageHandler[255];
+        private Type[] m_ReverseTypeMap = new Type[255];
+
+        private Dictionary<Type, byte> m_MessageTypes = new Dictionary<Type, byte>();
+        private Dictionary<ulong, NativeList<SendQueueItem>> m_SendQueues = new Dictionary<ulong, NativeList<SendQueueItem>>();
+
+        private List<INetworkHooks> m_Hooks = new List<INetworkHooks>();
```

Comment:
this hooks model is effectively replacing metrics and profiling stuff.
I think these proposed changes need a tools team pass — @Unity-Technologies/multiplayer-tools

Reply:
Took a look, This doesn't replace our metrics reporting, it just adjusts where the messaging system metrics callsites exist. I think the change looks good, and as long as the metrics reporting tests are still passing then we are good.

Reply:
I wasn't asking about the functionality breaking, I was asking about the new style/pattern. if things are OK, that's OK :)